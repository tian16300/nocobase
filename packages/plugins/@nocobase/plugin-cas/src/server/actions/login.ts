import { Context, Next } from '@nocobase/actions';
import { CASAuth } from '../auth';

export const login = async (ctx: Context, next: Next) => {
  const { authenticator } = ctx.action.params;
  const auth = (await ctx.app.authManager.get(authenticator, ctx)) as CASAuth;
  const { casUrl, serviceUrl } = auth.getOptions();
  const service = encodeURIComponent(`${serviceUrl}?authenticator=${authenticator}&__appName=${ctx.app.name}`);
  ctx.redirect(`${casUrl}/login?service=${service}`);
  next();
};
