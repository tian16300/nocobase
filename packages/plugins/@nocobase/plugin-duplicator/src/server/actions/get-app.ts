export async function getApp(ctx, subAppName) {
  let app = ctx.app;

  if (subAppName) {
    const subApp = await app.appManager.getApplication(subAppName);
    if (!subApp) {
      throw new Error(`app ${subAppName} not found`);
    }

    app = subApp;
  }

  return app;
}
