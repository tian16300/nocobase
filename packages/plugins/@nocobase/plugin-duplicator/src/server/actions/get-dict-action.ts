export default async function getDictAction(ctx, next) {
  ctx.withoutDataWrapping = true;
  let collectionNames = await ctx.db.getRepository('collections').find();
  collectionNames = collectionNames.map((item) => item.get('name'));
  const collections: any[] = [];
  for (const [name, collection] of ctx.db.collections) {
    const columns: any[] = [];
    for (const key in collection.model.rawAttributes) {
      if (Object.prototype.hasOwnProperty.call(collection.model.rawAttributes, key)) {
        const attribute = collection.model.rawAttributes[key];
        columns.push({
          realName: attribute.field,
          name: key,
        });
      }
    }
    const item = {
      name,
      tableName: collection.model.tableName,
      title: collection.options.title,
      namespace: collection.options.namespace,
      duplicator: collection.options.duplicator,
      schema: collection.options.schema || process.env.DB_SCHEMA,
      // columns,
    };
    if (!item.namespace && collectionNames.includes(name)) {
      item.namespace = 'collection-manager';
      if (!item.duplicator) {
        item.duplicator = 'optional';
      }
    }
    collections.push(item);
  }
  ctx.body = collections;
  await next();
}
