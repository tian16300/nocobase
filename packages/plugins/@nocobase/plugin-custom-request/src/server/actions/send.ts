import { Context, Next } from '@nocobase/actions';
import { parse } from '@nocobase/utils';

import axios from 'axios';
import CustomRequestPlugin from '../plugin';

const getHeaders = (headers: Record<string, any>) => {
  return Object.keys(headers).reduce((hds, key) => {
    if (key.toLocaleLowerCase().startsWith('x-')) {
      hds[key] = headers[key];
    }
    return hds;
  }, {});
};

const arrayToObject = (arr: { name: string; value: string }[]) => {
  return arr.reduce((acc, cur) => {
    acc[cur.name] = cur.value;
    return acc;
  }, {});
};

const omitNullAndUndefined = (obj: any) => {
  return Object.keys(obj).reduce((acc, cur) => {
    if (obj[cur] !== null && typeof obj[cur] !== 'undefined') {
      acc[cur] = obj[cur];
    }
    return acc;
  }, {});
};

export async function send(this: CustomRequestPlugin, ctx: Context, next: Next) {
  const { filterByTk, resourceName, values = {} } = ctx.action.params;
  const {
    currentRecord = {
      id: 0,
      appends: [],
      data: {},
    },
  } = values;

  // root role has all permissions
  if (ctx.state.currentRole !== 'root') {
    const crRepo = ctx.db.getRepository('customRequestsRoles');
    const hasRoles = await crRepo.find({
      filter: {
        customRequestKey: filterByTk,
      },
    });

    if (hasRoles.length) {
      if (!hasRoles.find((item) => item.roleName === ctx.state.currentRole)) {
        return ctx.throw(403, 'custom request no permission');
      }
    }
  }

  const repo = ctx.db.getRepository(resourceName);
  const requestConfig = await repo.findOne({
    filter: {
      key: filterByTk,
    },
  });

  if (!requestConfig) {
    ctx.throw(404, 'request config not found');
  }

  ctx.withoutDataWrapping = true;

  const { collectionName, url, headers = [], params = [], data = {}, ...options } = requestConfig.options;
  let currentRecordVariables = {};
  if (collectionName && typeof currentRecord.id !== 'undefined') {
    const recordRepo = ctx.db.getRepository(collectionName);
    currentRecordVariables = await recordRepo.findOne({
      filterByTk: currentRecord.id,
      appends: currentRecord.appends,
    });
  }

  const variables = {
    currentRecord: {
      ...currentRecordVariables,
      ...currentRecord.data,
    },
    currentUser: ctx.auth.user,
    currentTime: new Date().toISOString(),
  };

  const axiosRequestConfig = {
    baseURL: ctx.origin,
    ...options,
    url: parse(url)(variables),
    headers: {
      Authorization: 'Bearer ' + ctx.getBearerToken(),
      ...getHeaders(ctx.headers),
      ...omitNullAndUndefined(parse(arrayToObject(headers))(variables)),
    },
    params: parse(arrayToObject(params))(variables),
    data: parse(data)(variables),
  };

  const requestUrl = axios.getUri(axiosRequestConfig);
  this.logger.info(`custom-request:send:${filterByTk} request url ${requestUrl}`);
  this.logger.info(
    `custom-request:send:${filterByTk} request config ${JSON.stringify({
      ...axiosRequestConfig,
      headers: {
        ...axiosRequestConfig.headers,
        Authorization: null,
      },
    })}`,
  );

  try {
    ctx.body = await axios(axiosRequestConfig).then((res) => {
      this.logger.info(`custom-request:send:${filterByTk} success`);
      return res.data;
    });
  } catch (err) {
    if (axios.isAxiosError(err)) {
      ctx.status = err.response?.status || 500;
      ctx.body = err.response?.data || { message: err.message };
      this.logger.error(
        `custom-request:send:${filterByTk} error. status: ${ctx.status}, body: ${
          typeof ctx.body === 'string' ? ctx.body : JSON.stringify(ctx.body)
        }`,
      );
    } else {
      this.logger.error(`custom-request:send:${filterByTk} error. status: ${ctx.status}, message: ${err.message}`);
      ctx.throw(500, err?.message);
    }
  }

  return next();
}
