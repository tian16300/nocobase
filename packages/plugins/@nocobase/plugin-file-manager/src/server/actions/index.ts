import actions from '@nocobase/actions';
import { createMiddleware, destroyMiddleware } from './attachments';

export default function ({ app }) {
  app.resourcer.use(createMiddleware);
  app.resourcer.registerActionHandler('upload', actions.create);

  app.resourcer.use(destroyMiddleware);
}
