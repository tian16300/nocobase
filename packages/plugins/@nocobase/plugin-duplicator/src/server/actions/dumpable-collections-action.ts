import { Dumper } from '../dumper';
import { getApp } from './get-app';

export default async function dumpableCollections(ctx, next) {
  ctx.withoutDataWrapping = true;

  const app = await getApp(ctx, ctx.request.query.app);
  const dumper = new Dumper(app);

  ctx.body = await dumper.dumpableCollections();

  await next();
}
