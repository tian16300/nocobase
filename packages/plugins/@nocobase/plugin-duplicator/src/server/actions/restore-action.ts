import { Restorer } from '../restorer';
import * as os from 'os';
import path from 'path';
import { getApp } from './get-app';

export async function restoreAction(ctx, next) {
  const { restoreKey, selectedOptionalGroups, selectedUserCollections } = ctx.request.body;
  const appName = ctx.request.body.app;

  const tmpDir = os.tmpdir();
  const filePath = path.resolve(tmpDir, restoreKey);

  const app = await getApp(ctx, appName);

  const restorer = new Restorer(app, {
    backUpFilePath: filePath,
  });

  await restorer.restore({
    selectedOptionalGroupNames: selectedOptionalGroups,
    selectedUserCollections,
  });

  await next();
}

export const getPackageContent = async (ctx, next) => {
  const file = ctx.file;
  const fileName = file.filename;

  const restorer = new Restorer(ctx.app, {
    backUpFilePath: file.path,
  });

  const restoreMeta = await restorer.parseBackupFile();

  ctx.body = {
    key: fileName,
    meta: restoreMeta,
  };

  await next();
};
