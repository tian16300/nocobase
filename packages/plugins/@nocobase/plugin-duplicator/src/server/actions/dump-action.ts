import { Dumper } from '../dumper';
import send from 'koa-send';
import { getApp } from './get-app';

export default async function dumpAction(ctx, next) {
  const data = <
    {
      selectedOptionalGroupNames: string[];
      selectedUserCollections: string[];
      app?: string;
    }
  >ctx.request.body;

  const app = await getApp(ctx, data.app);

  const dumper = new Dumper(app);

  const { filePath, dirname } = await dumper.dump(data);

  await send(ctx, filePath.replace(dirname, ''), {
    root: dirname,
  });

  await next();
}
