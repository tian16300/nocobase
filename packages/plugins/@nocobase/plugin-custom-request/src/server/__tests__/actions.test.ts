import { Context } from '@nocobase/actions';
import Database, { Repository } from '@nocobase/database';
import { MockServer, mockServer, supertest } from '@nocobase/test';

describe('actions', () => {
  let app: MockServer;
  let db: Database;
  let repo: Repository;
  let agent: ReturnType<MockServer['agent']>;
  let resource: ReturnType<ReturnType<MockServer['agent']>['resource']>;

  beforeAll(async () => {
    app = mockServer({
      registerActions: true,
      acl: false,
      plugins: ['users', 'auth', 'acl', 'custom-request'],
    });

    await app.loadAndInstall({ clean: true });
    db = app.db;
    repo = db.getRepository('customRequests');
    agent = app.agent();
    resource = agent.resource('customRequests');
  });

  describe('send', () => {
    let params = null;
    beforeAll(async () => {
      app.resourcer.getResource('customRequests').addAction('test', (ctx: Context) => {
        params = ctx.action.params.values;
        return ctx.action.params.values;
      });
      await repo.create({
        values: {
          key: 'test',
          options: {
            url: '/customRequests:test',
            method: 'GET',
            data: {
              username: '{{ currentRecord.username }}',
            },
          },
        },
      });
    });

    test('basic', async () => {
      const res = await resource.send({
        filterByTk: 'test',
      });
      expect(res.status).toBe(200);
      expect(params).toMatchSnapshot();
    });

    test('currentRecord.data', async () => {
      const res = await resource.send({
        filterByTk: 'test',
        values: {
          currentRecord: {
            data: {
              username: 'testname',
            },
          },
        },
      });
      expect(res.status).toBe(200);
      expect(params).toMatchSnapshot();
    });
  });
});
