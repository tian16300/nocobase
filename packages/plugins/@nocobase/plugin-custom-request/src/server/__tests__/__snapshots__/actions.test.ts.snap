// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions send basic 1`] = `undefined`;

exports[`actions send currentRecord.data 1`] = `
{
  "username": "testname",
}
`;
